include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -414,7 +414,7 @@ struct __parallel_transform_scan_single_group_submitter
     template <typename _KernelName, typename _Policy, typename _InRng, typename _OutRng, typename _InitType,
               typename _BinaryOperation, typename _UnaryOp>
     static auto
-    __launch_dynamic_bounds_scan(const _Policy& __policy, _InRng __in_rng, _OutRng __out_rng, ::std::size_t __n,
+    __launch_dynamic_bounds_scan(const _Policy& __policy, _InRng&& __in_rng, _OutRng&& __out_rng, ::std::size_t __n,
```

Comment:
We are not forwarding in this function and as far as I can see, we cannot.. Does it make sense to accept the function argument by const lvalue ref? and by non-const lvalue ref for output range? This is also relevant for 468 line

Reply:
It's possible to change the input and output ranges to const refs and I don't believe it would make a semantic difference. But we cannot change the output range to a non-const lvalue ref because the ranges are being created and passed as an rvalue here: https://github.com/oneapi-src/oneDPL/blob/1854cb66f4bd141c664f9198ce526205d59cd8eb/include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h#L134

Reply:
How output range can be `const`?

Reply:
I agree that we likely overuse forwarding references, and not only in this place. However, I'm afraid if we take the output by lvalue reference here, we would need to do the same in the caller, and then possibly everywhere up the call stack. Do we want to? 
Or maybe the right questions would be: is there any significant downside of using forwarding references even in case the parameters are not forwarded down the stack? And if yes, what needs to be done to make it right?

Reply:
> However, I'm afraid if we take the output by lvalue reference here, we would need to do the same in the caller, and then possibly everywhere up the call stack. Do we want to?

It's not necessarily so. When we work with forwarding references the caller should forward that to the function it calls (if it's further unused on the caller side, of course). The callee is allowed to choose how they take the parameter so, it is not necessarily forwarding reference by any other (or better to say, "right") way. The same situation with execution policy. We forward it down to the stack but the backend takes it how it wills.

> Or maybe the right questions are: is there any significant downside of using forwarding references even in case the parameters are not forwarded down the stack? And if yes, what needs to be done to make it right?

I don't see any downsides except the code looks "suspicious". I would try to do forwarding as much as we can (of course, in correct way) and decide how to accept that on the level where this very argument is used. It should allow us to do not modify the upper-level (stack-wise) code when we want to change something in the code that uses ranges and/or execution policies

Reply:
@rarutyun Do you have any objections with keeping the input and output ranges as forwarding references?